### PR TITLE
Add Visformer implementation

### PR DESCRIPTION
**Pull request by aditya-agrawal-30502**
#28

### CODE REVIEW

File: tests/test_models.py [hunk 1]
```diff
@@ -391,3 +390,22 @@ def test_cct():
     f = model(img_3channels_224)
     assert f.shape == (4, 1000)
     del model
+
+
+def test_visformer():
+    model = MODEL_REGISTRY.get("Visformer_S")(224, 1000)
```
Comment (NeelayS) [inline tests/test_models.py:396]:
```suggestion

    model = MODEL_REGISTRY.get("Visformer_S")(224, 1000)
```

File: tests/test_models.py [hunk 2]
```diff
@@ -391,3 +390,22 @@ def test_cct():
...
+    out = model(img_3channels_224)
+    assert out.shape == (4, 1000)
+    del model
+    model = MODEL_REGISTRY.get("Visformer_Ti")(224, 1000)
```
Comment (NeelayS) [inline tests/test_models.py:400]:
```suggestion

    model = MODEL_REGISTRY.get("Visformer_Ti")(224, 1000)
```

File: tests/test_models.py [hunk 3]
```diff
@@ -391,3 +390,22 @@ def test_cct():
...
+    out = model(img_3channels_224)
+    assert out.shape == (4, 1000)
+    del model
+    model = MODEL_REGISTRY.get("VisformerV2_S")(224, 1000)
```
Comment (NeelayS) [inline tests/test_models.py:404]:
```suggestion

    model = MODEL_REGISTRY.get("VisformerV2_S")(224, 1000)
```

File: tests/test_models.py [hunk 4]
```diff
@@ -391,3 +390,22 @@ def test_cct():
...
+    out = model(img_3channels_224)
+    assert out.shape == (4, 1000)
+    del model
+    model = MODEL_REGISTRY.get("VisformerV2_Ti")(224, 1000)
```
Comment (NeelayS) [inline tests/test_models.py:408]:
```suggestion

    model = MODEL_REGISTRY.get("VisformerV2_Ti")(224, 1000)
```

File: vformer/models/classification/visformer.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+import torch.nn as nn
+
+from vformer.attention import VanillaSelfAttention
+
+from ...utils import MODEL_REGISTRY
+
+
+# need to add visformerV2_ti
+# editted number of need to change
+# run one epoch to check if the model is working
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:11]: Please remove these comments once you are done.

File: vformer/models/classification/visformer.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+class Conv_Block(nn.Module):
+    def __init__(self, in_channels, group=8, activation=nn.GELU, drop=0.0):
+        super(Conv_Block, self).__init__()
+        self.norm1 = nn.BatchNorm2d(in_channels)
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:15]:
```suggestion

        self.norm1 = nn.BatchNorm2d(in_channels)
```

File: vformer/models/classification/visformer.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        xt = x
+        xt = self.norm1(xt)
+        x = self.conv1(x)
+        x = self.act1(x)
+        x = self.drop(x)
+        x = self.conv2(x)
+        x = self.act2(x)
+        x = self.conv3(x)
+        x = xt + self.drop(x)
+        del xt
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:40]: Don't need to do this.

File: vformer/models/classification/visformer.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            padding=1,
+            groups=group,
+            bias=False,
+        )
+        self.act2 = activation()
+        self.conv3 = nn.Conv2d(in_channels * 2, in_channels, kernel_size=1, bias=False)
+        self.drop = nn.Dropout(drop)
+
+    def forward(self, x):
+        xt = x
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:31]:
```suggestion

        xt = x
```

File: vformer/models/classification/visformer.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        xt = self.norm1(xt)
+        x = self.conv1(x)
+        x = self.act1(x)
+        x = self.drop(x)
+        x = self.conv2(x)
+        x = self.act2(x)
+        x = self.conv3(x)
+        x = xt + self.drop(x)
+        del xt
+        return x
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:41]:
```suggestion

        return x
```

File: vformer/models/classification/visformer.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+
+class Attention_Block(nn.Module):
+    def __init__(self, channels, num_heads=8, drop=0.0):
+        super().__init__()
+        self.conv1 = nn.Conv2d(channels, channels * 4, kernel_size=1, bias=False)
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:47]:
```suggestion

        self.conv1 = nn.Conv2d(channels, channels * 4, kernel_size=1, bias=False)
```

File: vformer/models/classification/visformer.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+        x = self.attn(x)
+        x = xt + x
+        x = einops.rearrange(x, "b (h w) c -> b c h w", h=H, w=W)
+        xt = x
+        x = self.norm2(x)
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = self.drop(x)
+        x = xt + x
+        del xt
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:70]: Don't need to do this.

File: vformer/models/classification/visformer.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        self.conv2 = nn.Conv2d(channels * 4, channels, kernel_size=1, bias=False)
+        self.attn = VanillaSelfAttention(
+            channels, num_heads=num_heads, head_dim=channels // num_heads
+        )
+        self.norm1 = nn.BatchNorm2d(channels)
+        self.norm2 = nn.BatchNorm2d(channels)
+        self.drop = nn.Dropout(drop)
+
+    def forward(self, x):
+        B, C, H, W = x.shape
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:57]:
```suggestion 

        B, C, H, W = x.shape
```

File: vformer/models/classification/visformer.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        x = xt + x
+        x = einops.rearrange(x, "b (h w) c -> b c h w", h=H, w=W)
+        xt = x
+        x = self.norm2(x)
+        x = self.conv1(x)
+        x = self.conv2(x)
+        x = self.drop(x)
+        x = xt + x
+        del xt
+        return x
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:71]:
```suggestion 

        return x
```

File: vformer/models/classification/visformer.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+        image_size,
+        n_classes,
+        depth: list,
+        config: str,
+        channel_config: list,
+        num_heads=8,
+        drop=0.0,
+    ):
+        super().__init__()
+        q = 0
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:87]:
```suggestion 

        q = 0
```

File: vformer/models/classification/visformer.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+class Visformer(nn.Module):
+    def __init__(
+        self,
+        image_size,
+        n_classes,
+        depth: list,
+        config: str,
+        channel_config: list,
+        num_heads=8,
+        drop=0.0,
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:84]: @abhi-glitchhg had made some changes where he had standardized the names of parameters like dropout across all files. Could you check if this is consistent with those names?

**Review by NeelayS — CHANGES_REQUESTED** [review]
I have added a few comments, please check.

File: vformer/models/classification/visformer.py (hunk 15 above)
Comment (aditya-agrawal-30502) [reply vformer/models/classification/visformer.py:84]: alright lemme see

File: vformer/models/classification/visformer.py (hunk 11 above)
Comment (aditya-agrawal-30502) [reply vformer/models/classification/visformer.py:109]: do you mean deleting xt?
just did it to be sure

File: vformer/models/classification/visformer.py [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+    pos_embedding: bool
+        Whether to use positional embedding, default is True
+
+    """
+
+    def __init__(
+        self,
+        img_size,
+        n_classes,
+        depth: tuple,
```
Comment (abhi-glitchhg) [inline vformer/models/classification/visformer.py:151]:
```suggestion
    def __init__(
        self,
        img_size,
        num_classes,
        depth: tuple,
```
As per the docstring

Comment (aditya-agrawal-30502) [reply vformer/models/classification/visformer.py:151]: changes the docstring to maintain common naming in package

**Comment by abhi-glitchhg** [comment]
fixes #28

**Review by abhi-glitchhg — COMMENTED** [review]
👍 
Great work
Seems good according to me,  just have a look on the comment I left.

File: vformer/models/classification/visformer.py [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+        x = x + xt
+        del xt
+
+        return x
+
+
+@MODEL_REGISTRY.register()
+class Visformer_Attention_Block(nn.Module):
+    """
+    Attention Block for Vision-Friendly transformers
```
[the remark this answers is not in the document]
Comment (aditya-agrawal-30502) [reply vformer/models/classification/visformer.py:65]: i am not quite sure where to put it as its not exactly a new attention head so it wouldnt really fit in the attention registry.
to me it seems more like a small model itself so should be in the model registry

File: vformer/models/classification/visformer.py [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+        self.stem.extend([nn.BatchNorm2d(channel_config[-1]), nn.AdaptiveAvgPool2d(1)])
+        self.softmax = nn.Softmax(dim=1)
+
+    def forward(self, x):
+        for i in self.stem:
+            x = i(x)
+        x.squeeze_(2).squeeze_(2)
+        x = self.linear(x)
+        x = self.softmax(x)
+        return x
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:250]:
```suggestion

        return x
```

File: vformer/models/classification/visformer.py [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+                        for j in range(depth[i])
+                    ]
+                )
+        self.stem.extend([nn.BatchNorm2d(channel_config[-1]), nn.AdaptiveAvgPool2d(1)])
+        self.softmax = nn.Softmax(dim=1)
+
+    def forward(self, x):
+        for i in self.stem:
+            x = i(x)
+        x.squeeze_(2).squeeze_(2)
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:247]:
```suggestion

        x.squeeze_(2).squeeze_(2)
```

**Review by NeelayS — CHANGES_REQUESTED** [review]

File: vformer/models/classification/visformer.py [hunk 20]
```diff
@@ -0,0 +1,365 @@
+import einops
+import torch.nn as nn
+
+from ...attention import VanillaSelfAttention
+from ...encoder.embedding.pos_embedding import PosEmbedding
+from ...utils import MODEL_REGISTRY, ATTENTION_REGISTRY
```
Comment (NeelayS) [inline vformer/models/classification/visformer.py:6]:
```suggestion
from ...utils import ATTENTION_REGISTRY, MODEL_REGISTRY
```
